pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -365,11 +366,11 @@ func (m *managerImpl) processShutdownEvent() error {
 			go func(pod *v1.Pod, group podShutdownGroup) {
 				defer wg.Done()
 
-				gracePeriodOverride := group.ShutdownGracePeriodSeconds
+				gracePeriodOverride := int64(v1.DefaultTerminationGracePeriodSeconds)
```

Comment:
A question regarding also `Test pods and many other pods do not set a termination grace period and so would be set to potentially very high timeouts (65 minutes is what I am configuring). `: what is the value of `.spec.terminationGracePeriodSeconds` for such pods actually once created (after defaulting)?

I'm wondering since we have the defaulting configured here: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go#L240-L243. So, IIUC pods have the field set to `DefaultTerminationGracePeriodSeconds` 30s (if not specified otherwise).

Reply:
The defaulting does not take into account the user defined spec field. We cannot rely on the defaulting to see what the graceful shutdown field is set to.

Reply:
Can you explain a bit more? IIUC `.spec.TerminationGracePeriodSeconds` is the field that can be set by the user. If it is not set by the user, then `defaults.go` sets it to 30s. So, IIUC the shutdown manager code would always see `.spec.TerminationGracePeriodSeconds` as set (either by the user or defaulter).

Reply:
.spec.TerminationGracePeriodSeconds isn't explicitly set. The code you linked only defaults .TerminationGracePeriodSeconds (not on the spec field).

Reply:
The `obj` in the referenced code is `*v1.PodSpec`, so I assumed this is spec. When I create a pod without the field in yaml, the created pod has it, when I do `kubectl get pods -oyaml`:
```yaml
  spec:
    containers:
    - args:
      - -c
      - |
        echo "Hello world!"
    ...
    terminationGracePeriodSeconds: 30
```
This is I would like to better understand what I'm missing and how you create the pods that don't have `.spec.terminationGracePeriodSeconds` set.

Reply:
pod_worker and other areas of the kubelet always check for a nil terminationGracePeriodSeconds. Are you suggesting we should not do a nil check?

Reply:
I think if other places in the code check this then indeed it is probably needed. However I wanted to ask / investigate how this is possible (from the description): `This fixes a bug when Node Graceful Shutdown is enabled where all pods get a terminationGracePeriod of the maximum time set for graceful shutdown.`.

Reply:
@bobbypage do you have any insights here?

Reply:
How about static pods. I'm not sure,  but might be they don't go via the defaulting. In that case the check is needed, but it still does not explain why the pods created by API might get the grace period arbitrarily large.

Reply:
Yes, but assuming the pod's `.spec.terminationGracePeriod` is defaulted to 30s it would override that value here: 
https://github.com/kubernetes/kubernetes/blob/a759bf2b1e5c4964f6850d76aafafd5ae985442b/pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go#L372

Reply:
We are specifically seeing this on E2E tests, since we are running the e2e tests with graceful shutdown enabled. I'm thinking the e2e tests are using the raw structure (without the defaulting).

Reply:
I would expect that as long as the e2e tests create the pods via API Server, the field should be defaulted.

Reply:
Taking a look into this /hold

Reply:
/hold